projects/Mallard/src/hooks/use-cached.ts

```diff
@@ -0,0 +1,63 @@
+import NetInfo from '@react-native-community/netinfo'
+
+import { useEffect, useState } from 'react'
+import { useSettings } from './use-settings'
+import axios from 'axios'
+import RNFetchBlob from 'rn-fetch-blob'
+import { DatePickerAndroid } from 'react-native'
+
+export const useCached = <Fetchable>(
+    path: string,
+    initialState: Fetchable,
+): Fetchable => {
+    const [{ apiUrl }] = useSettings()
+    const localPath = `storage/${path}`
+
+    const [data, updateData] = useState(initialState)
+
+    const fetchAndStore = async (path: string, date: Date | null = null) => {
+        const network = await NetInfo.fetch()
+        const headers = date
+            ? {
+                  'If-Modified-Since': date.toUTCString(),
+              }
+            : {}
+        if (!network.isConnected) {
+            //do something to deal with this here!
+            return
+        }
+        const resp = await axios(`${apiUrl}/${path}`, { headers })
+        if (resp.status == 304) {
+            return
+        }
+        const data = resp.data
+        const modified = resp.headers['Last-Modified']
+        RNFetchBlob.fs.createFile(localPath, data, 'utf8')
+        RNFetchBlob.fs.createFile(
+            `${localPath}.date`,
+            JSON.stringify({ date: modified }),
+            'utf8',
+        )
+    }
+    useEffect(() => {
```

Comment:
can we keep the hook as small as possible? we can probably turn `fetchWithCache()` into a normal js promise-like operation and then the hook just interfaces with that and it's a bit more separate